# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLProgram.h
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 dom/canvas/WebGLProgram.h
--- a/dom/canvas/WebGLProgram.h	Wed Jan 08 21:52:03 2020 +0000
+++ b/dom/canvas/WebGLProgram.h	Wed Jan 08 22:19:14 2020 +0000
@@ -18,12 +18,14 @@
 #include "nsWrapperCache.h"
 
 #include "CacheInvalidator.h"
+#include "WebGLActiveInfo.h"
 #include "WebGLContext.h"
 #include "WebGLObjectModel.h"
 
 namespace mozilla {
 class ErrorResult;
 class WebGLActiveInfo;
+class WebGLContext;
 class WebGLProgram;
 class WebGLShader;
 class WebGLUniformLocation;
@@ -41,14 +43,14 @@
 enum class TextureBaseType : uint8_t;
 
 struct AttribInfo final {
-  const RefPtr<WebGLActiveInfo> mActiveInfo;
+  const WebGLActiveInfo mActiveInfo;
   const GLint mLoc;  // -1 for active built-ins
 };
 
 struct UniformInfo final {
   typedef decltype(WebGLContext::mBound2DTextures) TexListT;
 
-  const RefPtr<WebGLActiveInfo> mActiveInfo;
+  const WebGLActiveInfo mActiveInfo;
   const TexListT* const mSamplerTexList;
   const webgl::TextureBaseType mTexBaseType;
   const bool mIsShadowSampler;
@@ -56,10 +58,11 @@
   std::vector<uint32_t> mSamplerValues;
 
  protected:
-  static const TexListT* GetTexList(WebGLActiveInfo* activeInfo);
+  static const TexListT* GetTexList(const WebGLContext* aWebGL,
+                                    WebGLActiveInfo* activeInfo);
 
  public:
-  explicit UniformInfo(WebGLActiveInfo* activeInfo);
+  explicit UniformInfo(const WebGLContext* aWebGL, WebGLActiveInfo& activeInfo);
 };
 
 struct UniformBlockInfo final {
@@ -106,7 +109,7 @@
   std::vector<AttribInfo> attribs;
   std::vector<UniformInfo*> uniforms;            // Owns its contents.
   std::vector<UniformBlockInfo*> uniformBlocks;  // Owns its contents.
-  std::vector<RefPtr<WebGLActiveInfo>> transformFeedbackVaryings;
+  std::vector<WebGLActiveInfo> transformFeedbackVaryings;
   std::unordered_map<uint8_t, const FragOutputInfo> fragOutputs;
   uint8_t zLayerCount = 1;
 
@@ -138,15 +141,13 @@
 
 }  // namespace webgl
 
-class WebGLProgram final : public nsWrapperCache,
-                           public WebGLRefCountedObject<WebGLProgram>,
+class WebGLProgram final : public WebGLRefCountedObject<WebGLProgram>,
                            public LinkedListElement<WebGLProgram> {
   friend class WebGLTransformFeedback;
   friend struct webgl::LinkedProgramInfo;
 
  public:
-  NS_INLINE_DECL_CYCLE_COLLECTING_NATIVE_REFCOUNTING(WebGLProgram)
-  NS_DECL_CYCLE_COLLECTION_SCRIPT_HOLDER_NATIVE_CLASS(WebGLProgram)
+  NS_INLINE_DECL_REFCOUNTING(WebGLProgram)
 
   explicit WebGLProgram(WebGLContext* webgl);
 
@@ -156,25 +157,23 @@
   void AttachShader(WebGLShader* shader);
   void BindAttribLocation(GLuint index, const nsAString& name);
   void DetachShader(const WebGLShader* shader);
-  already_AddRefed<WebGLActiveInfo> GetActiveAttrib(GLuint index) const;
-  already_AddRefed<WebGLActiveInfo> GetActiveUniform(GLuint index) const;
-  void GetAttachedShaders(nsTArray<RefPtr<WebGLShader>>* const out) const;
+  Maybe<WebGLActiveInfo> GetActiveAttrib(GLuint index) const;
+  Maybe<WebGLActiveInfo> GetActiveUniform(GLuint index) const;
+  MaybeAttachedShaders GetAttachedShaders() const;
   GLint GetAttribLocation(const nsAString& name) const;
   GLint GetFragDataLocation(const nsAString& name) const;
-  void GetProgramInfoLog(nsAString* const out) const;
-  JS::Value GetProgramParameter(GLenum pname) const;
+  nsString GetProgramInfoLog() const;
+  MaybeWebGLVariant GetProgramParameter(GLenum pname) const;
   GLuint GetUniformBlockIndex(const nsAString& name) const;
-  void GetActiveUniformBlockName(GLuint uniformBlockIndex,
-                                 nsAString& name) const;
-  JS::Value GetActiveUniformBlockParam(GLuint uniformBlockIndex,
-                                       GLenum pname) const;
-  JS::Value GetActiveUniformBlockActiveUniforms(
-      JSContext* cx, GLuint uniformBlockIndex,
-      ErrorResult* const out_error) const;
+  nsString GetActiveUniformBlockName(GLuint uniformBlockIndex) const;
+  MaybeWebGLVariant GetActiveUniformBlockParam(GLuint uniformBlockIndex,
+                                               GLenum pname) const;
+  MaybeWebGLVariant GetActiveUniformBlockActiveUniforms(
+      GLuint uniformBlockIndex) const;
   already_AddRefed<WebGLUniformLocation> GetUniformLocation(
       const nsAString& name) const;
-  void GetUniformIndices(const dom::Sequence<nsString>& uniformNames,
-                         dom::Nullable<nsTArray<GLuint>>& retval) const;
+  MaybeWebGLVariant GetUniformIndices(
+      const nsTArray<nsString>& uniformNames) const;
   void UniformBlockBinding(GLuint uniformBlockIndex,
                            GLuint uniformBlockBinding) const;
 
@@ -195,10 +194,9 @@
   bool UnmapUniformBlockName(const nsCString& mappedName,
                              nsCString* const out_userName) const;
 
-  void TransformFeedbackVaryings(const dom::Sequence<nsString>& varyings,
+  void TransformFeedbackVaryings(const nsTArray<nsString>& varyings,
                                  GLenum bufferMode);
-  already_AddRefed<WebGLActiveInfo> GetTransformFeedbackVarying(
-      GLuint index) const;
+  Maybe<WebGLActiveInfo> GetTransformFeedbackVarying(GLuint index) const;
 
   void EnumerateFragOutputs(
       std::map<nsCString, const nsCString>& out_FragOutputs) const;
@@ -212,11 +210,6 @@
   const auto& VertShader() const { return mVertShader; }
   const auto& FragShader() const { return mFragShader; }
 
-  WebGLContext* GetParentObject() const { return mContext; }
-
-  virtual JSObject* WrapObject(JSContext* js,
-                               JS::Handle<JSObject*> givenProto) override;
-
  private:
   ~WebGLProgram();
 